flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/AbstractMetricGroup.java

```diff
@@ -71,7 +73,7 @@
 	protected final A parent;
 
 	/** The map containing all variables and their associated values, lazily computed. */
-	protected volatile Map<String, String> variables;
+	protected volatile Map<String, String>[] variables;
```

Comment:
`variables`, `scopeComponents`, `scopeStrings` and `logicalScopeStrings` seem to belong together. It might make sense to group them and to pull them out of the `AbstractMetricGroup` in order to reduce the responsibilities of this class. We should do this as a follow up, though.